hail/python/hail/backend/service_backend.py

```diff
@@ -1,157 +1,370 @@
+                else:
+                    jstacktrace = await read_str(outfile)
+                    maybe_id = ServiceBackend.HAIL_BATCH_FAILURE_EXCEPTION_MESSAGE_RE.match(jstacktrace)
+                    if maybe_id:
+                        batch_id = maybe_id.groups()[0]
+                        b2 = await self.async_bc.get_batch(batch_id)
+                        b2_status = await b2.status()
+                        assert b2_status['state'] != 'success'
+                        failed_jobs = []
+                        async for j in b2.jobs():
+                            if j['state'] != 'Success':
+                                logs, job = await asyncio.gather(
+                                    self.async_bc.get_job_log(j['batch_id'], j['job_id']),
+                                    self.async_bc.get_job(j['batch_id'], j['job_id']),
+                                )
+                                full_status = job._status
+                                if 'status' in full_status:
+                                    if 'error' in full_status['status']:
+                                        full_status['status']['error'] = yaml_literally_shown_str(full_status['status']['error'].strip())
+                                main_log = logs.get('main', '')
+                                failed_jobs.append({
+                                    'partial_status': j,
+                                    'status': full_status,
+                                    'log': yaml_literally_shown_str(main_log.strip()),
+                                })
+                        message = {
+                            'id': b.id,
+                            'stacktrace': yaml_literally_shown_str(jstacktrace.strip()),
+                            'cause': {'id': batch_id, 'batch_status': b2_status, 'failed_jobs': failed_jobs}}
+                        log.error(yaml.dump(message))
+                        raise ValueError(json.dumps(message))
+                    raise FatalError(f'batch id was {b.id}\n' + jstacktrace)
+
+    async def _async_execute_untimed(self, ir):
+        async def inputs(infile, token):
+            await write_int(infile, ServiceBackend.EXECUTE)
+            await write_str(infile, tmp_dir())
+            await write_str(infile, self.billing_project)
+            await write_str(infile, self.remote_tmpdir)
+            await write_str(infile, self.render(ir))
+            await write_str(infile, token)
+        _, resp = await self._rpc('execute(...)', inputs)
         typ = dtype(resp['type'])
         if typ == tvoid:
-            value = None
-        else:
-            value = typ._convert_from_json_na(resp['value'])
-        # FIXME put back timings
+            return None
+        return typ._convert_from_json_na(resp['value'])
 
-        return (value, None) if timed else value
+    def execute_many(self, *irs, timed=False):
```

Comment:
Could wrap in a tuple and submit a single query, but I think this is really just used for the tests, right? If so then we should leave as-is to give each query the authentic isolation experience.

Reply:
Yes and agreed. This is really only useful for backends that can run queries in parallel.